bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java

```diff
@@ -0,0 +1,433 @@
+        } catch (MalformedURLException e) {
+            throw new MiCloudException("Error getting logon URL. Cannot logon to Xiaomi cloud: " + e.getMessage(), e);
+        }
+    }
+
+    private String loginStep1() throws InterruptedException, TimeoutException, ExecutionException, MiCloudException {
+        final ContentResponse responseStep1;
+
+        logger.trace("Xiaomi Login step 1");
+        String url = "https://account.xiaomi.com/pass/serviceLogin?sid=xiaomiio&_json=true";
+        Request request = httpClient.newRequest(url).timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        request.agent(USERAGENT);
+        request.header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded");
+        request.cookie(new HttpCookie("userId", this.userId.length() > 0 ? this.userId : this.username));
+
+        responseStep1 = request.send();
+        final String content = responseStep1.getContentAsString();
+        logger.trace("Xiaomi Login step 1 content response= {}", content);
+        logger.trace("Xiaomi Login step 1 response = {}", responseStep1);
+        try {
+            JsonElement resp = new JsonParser().parse(parseJson(content));
+            String sign = resp.getAsJsonObject().get("_sign").getAsString();
+            logger.trace("Xiaomi Login step 1 sign = {}", sign);
+            return sign;
+        } catch (JsonSyntaxException | NullPointerException e) {
+            throw new MiCloudException("Error getting logon sign. Cannot parse response: " + e.getMessage(), e);
+        }
+    }
+
+    private String loginStep2(String sign)
+            throws MiIoCryptoException, InterruptedException, TimeoutException, ExecutionException, MiCloudException {
+        String passToken;
+        String cUserId;
+
+        logger.trace("Xiaomi Login step 2");
+        String url = "https://account.xiaomi.com/pass/serviceLoginAuth2";
+        Request request = httpClient.newRequest(url).timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        request.agent(USERAGENT);
+        request.method(HttpMethod.POST);
+        final ContentResponse responseStep2;
+
+        Fields fields = new Fields();
+        fields.put("sid", "xiaomiio");
+        // fields.put("hash", encodePassword(password));
+        fields.put("hash", Utils.getHex(MiIoCrypto.md5(password.getBytes())));
+        fields.put("callback", "https://sts.api.io.mi.com/sts");
+        fields.put("qs", "%3Fsid%3Dxiaomiio%26_json%3Dtrue");
+        fields.put("user", username);
+        fields.put("_sign", sign);
+        fields.put("_json", "true");
+
+        request.content(new FormContentProvider(fields));
+        responseStep2 = request.send();
+
+        final String content2 = responseStep2.getContentAsString();
+        logger.trace("Xiaomi login step 2 response = {}", responseStep2);
+        logger.trace("Xiaomi login step 2 content = {}", content2);
+
+        JsonElement resp2 = new JsonParser().parse(parseJson(content2));
```

Comment:
Instead dealing with JsonElement, wouldn't it be better to parse json into a DTO instead?

Reply:
Yes, did that, as well as with the device response. That actually improved the code in several places.